### DISCUSSION TITLE
Undefined symbol PsrHttpMessageUriInterface_ce_ptr with PHP8.0, but PHP7.4 works... (Setting up Phalcon 5.0.0-alpha with Docker Compose)

### DISCUSSION THREAD

**Discussion by aircodepl** [phalcon/cphalcon#15535] [Q&A] [answered]
Hello!

I cannot load Phalcon 5.0.0-alpha.2 with PHP 8.0. It only works with PHP 7.4.

Here is a warning from PHP Startup:

```
Warning: PHP Startup: Unable to load dynamic library 'phalcon.so' (tried: /usr/local/lib/php/extensions/no-debug-non-zts-20200930/phalcon.so (/usr/local/lib/php/extensions/no-debug-non-zts-20200930/phalcon.so: undefined symbol: PsrHttpMessageUriInterface_ce_ptr), /usr/local/lib/php/extensions/no-debug-non-zts-20200930/phalcon.so.so (/usr/local/lib/php/extensions/no-debug-non-zts-20200930/phalcon.so.so: cannot open shared object file: No such file or directory)) in Unknown on line 0
```

Ini is being parsed, but extension not loaded.

I'm basing on this script: https://github.com/MilesChou/docker-phalcon

```
FROM php:8.0-apache

LABEL maintainer="MilesChou <github.com/MilesChou>, fizzka <github.com/fizzka>"

ARG PSR_VERSION=1.1.0
ARG PHALCON_VERSION=5.0.0-alpha.2
ARG PHALCON_EXT_PATH=phalcon/64bits

RUN set -xe && \
        # Download PSR, see https://github.com/jbboehr/php-psr
        curl -LO https://github.com/jbboehr/php-psr/archive/v${PSR_VERSION}.tar.gz && \
        tar xzf ${PWD}/v${PSR_VERSION}.tar.gz && \
        # Download Phalcon
        curl -LO https://github.com/phalcon/cphalcon/archive/refs/tags/v${PHALCON_VERSION}.tar.gz && \
        tar xzf ${PWD}/v${PHALCON_VERSION}.tar.gz && \
        docker-php-ext-install -j $(getconf _NPROCESSORS_ONLN) \
            ${PWD}/php-psr-${PSR_VERSION} \
            ${PWD}/cphalcon-${PHALCON_VERSION}/build/${PHALCON_EXT_PATH} \
        && \
        # Remove all temp files
        rm -r \
            ${PWD}/v${PSR_VERSION}.tar.gz \
            ${PWD}/php-psr-${PSR_VERSION} \
            ${PWD}/v${PHALCON_VERSION}.tar.gz \
            ${PWD}/cphalcon-${PHALCON_VERSION} \
        && \
        php -m

COPY docker-phalcon-* /usr/local/bin/
```

Other extensions works well.

Also, when I change image to php:7.4-apache, Phalcon runs without problems.

**Comment by kaioken** [comment]
@aircodepl did you get this solved?

**Comment by kaioken** [comment]
@aircodepl found the issue , 

```

**Comment by kaioken** [comment]
@aircodepl solved the issue, you just need to 

The undefined symbol PSR..... is because PSR is loaded after Phalcon. Try to put Phalcon below PSR. That will make it work.

thanks to @niden

**Comment by niden** [comment]
For those that use scrutinizer, installing PSR and Phalcon from PECL will put Phalcon above PSR in the `php.ini`. The solution for your `scrutinizer.yml` file is:

```
# We need to remove phalcon.so from php.ini and add it after psr.so
- sed -i 's/extension="phalcon.so"/;/g' /home/scrutinizer/.phpenv/versions/8.0.10/etc/php.ini
- sed -i 's/extension="psr.so"/extension="psr.so"\nextension="phalcon.so"\n/g' /home/scrutinizer/.phpenv/versions/8.0.10/etc/php.ini
```

The above can be adjusted to suit your needs when setting up your CI or docker environment. For bare bones server just look at the load order.

**Comment by aircodepl** [comment, marked as answer]
Hi and I'm really sorry - I thought that I removed this topic after I realized it's because of PSR being loaded after Phalcon.

Yes, week or two ago I did manage to install Phalcon 5 with Dockerimage and did it with similar to niden (but pathetic) trick:

```
# Rename config file
RUN mv '/usr/local/etc/php/conf.d/docker-php-ext-psr.ini' '/usr/local/etc/php/conf.d/docker-php-ext-phalcon-psr.ini'
```

I renamed ini file of PSR to be alphabetically loaded before. lol

Here is the whole Dockerfile script:
https://github.com/aircodepl/phalcon-5-dockerfile/blob/master/Dockerfile

Of course, if anyone want to check it out, you need to keep in mind that I suck at setting up environments, especially docker. Dockerfiles that I managed to create are the only ones that I was able to run with just one simple command. Any other that I found over the internet or in Phalcon repository was getting me errors and all sort of problems. These that I wrote after looking up at MilesChou examples are working with no problem. _docker-compose up -d_ and I'm ready to work.
